Pinta.Core/Classes/AsyncEffectRenderer.cs

```diff
@@ -39,15 +39,38 @@ namespace Pinta.Core;
 // Only call methods on this class from a single thread (The UI thread).
 internal abstract class AsyncEffectRenderer
 {
-	Settings settings;
+	private readonly Settings settings;
 
-	internal struct Settings
+	internal sealed class Settings
 	{
-		internal int ThreadCount { get; set; }
-		internal int TileWidth { get; set; }
-		internal int TileHeight { get; set; }
-		internal int UpdateMillis { get; set; }
-		internal ThreadPriority ThreadPriority { get; set; }
+		internal int ThreadCount { get; }
+		internal int TileWidth { get; }
+		internal int TileHeight { get; }
+		internal int UpdateMillis { get; }
+		internal ThreadPriority ThreadPriority { get; }
+
+		internal Settings (
+			int threadCount,
+			int tileWidth,
+			int tileHeight,
+			int updateMilliseconds,
+			ThreadPriority threadPriority)
+		{
+			if (tileWidth < 0)
+				throw new ArgumentOutOfRangeException (nameof (tileWidth), "Cannot be negative");
+
+			if (tileHeight < 0)
+				throw new ArgumentOutOfRangeException (nameof (tileHeight), "Cannot be negative");
+
+			// TODO: Prepare the calling code and
+			//       instead of coercing values,
+			//       throw an exception
+			TileWidth = tileWidth;
+			TileHeight = tileHeight;
+			ThreadCount = Math.Clamp (threadCount, 1, int.MaxValue);
```

Comment:
This could just be `Math.Max()` since clamping to `int.MaxValue` is redundant.
But I'm not sure this line is really necessary either since `SystemManager.RenderThreads` is never less than 1

Reply:
These lines were more or less a literal translation from mutable to immutable, but it is true that I ended up with questionable code.

I have now changed the coercions to exception throwing. We shouldn't expect these exceptions to actually be thrown, as the values passed are always within a valid range (additionally, I tested it on an Ubuntu VM just to make sure).